src/main.cpp

```diff
@@ -0,0 +1,1605 @@
+    // differ from window resolution.
+    if (capabilities.currentExtent.width != std::numeric_limits<uint32_t>::max()) {
+        return capabilities.currentExtent;
+    }
+
+    SDL_Vulkan_GetDrawableSize(
+            window,
+            &pixels_drawable_width,
+            &pixels_drawable_height);
+
+    extent = {
+        static_cast<uint32_t>(pixels_drawable_width),
+        static_cast<uint32_t>(pixels_drawable_height)
+    };
+
+    extent.width = std::clamp(
+            extent.width,
+            capabilities.minImageExtent.width,
+            capabilities.maxImageExtent.width);
+    extent.height = std::clamp(
+            extent.height,
+            capabilities.minImageExtent.height,
+            capabilities.maxImageExtent.height);
+
+    return extent;
+}
+
+static VkSwapchainKHR
+create_swap_chain(
+        const VkPhysicalDevice& device,
+        const VkDevice& logical_device,
+        const VkSurfaceKHR& surface,
+        SDL_Window* window,
+        VkFormat& swapchain_image_format,
+        VkExtent2D& swapchain_extent)
+{
+    bool condition;
+    std::uint32_t image_count;
+
+    std::uint32_t family_index_count;
+    const std::uint32_t* family_index_array_base;
+
+    struct QueueFamilyIndices family_indices;
+    struct SurfaceProperties surface_properties;
+
+    VkResult result;
+
+    VkSurfaceFormatKHR format;
+    VkPresentModeKHR present_mode;
+    VkExtent2D extent;
+
+    VkSharingMode sharing_mode;
+
+    VkSwapchainKHR swapchain;
+    VkSwapchainCreateInfoKHR chain_create_info;
+
+    condition = false;
+    image_count = 0;
+
+    sharing_mode = VK_SHARING_MODE_MAX_ENUM;
+
+    family_index_count = 0;
+
+    result = VK_ERROR_UNKNOWN;
+    family_indices = {0};
+    surface_properties = {0};
+    format = {};
+    present_mode = {};
+    extent = {};
+    swapchain = {};
+    chain_create_info = {};
+
+    std::uint32_t family_index_array[2] = {
+        family_indices.drawing_family,
+        family_indices.presentation_family,
+    };
+
+    surface_properties = find_surface_properties(device, surface);
+    format = pick_swap_surface_format(surface_properties.formats);
+    present_mode = pick_swap_present_mode(surface_properties.present_modes);
+    extent = pick_swap_extent(surface_properties.capabilities, window);
+    family_indices = find_queue_family_indices(device, surface);
+
+    sharing_mode = VK_SHARING_MODE_EXCLUSIVE;
+    family_index_count = 0;
+    family_index_array_base = nullptr;
+    condition =
+        family_indices.drawing_family !=
+        family_indices.presentation_family;
+    if (condition) {
+        // If drawing and presentation have separate queues, then using
+        // concurrent mode to avoid writing image ownership transfer code.
+        sharing_mode = VK_SHARING_MODE_CONCURRENT;
+        family_index_count = 2;
+        family_index_array_base = family_index_array;
+    }
+
+    image_count = surface_properties.capabilities.minImageCount + 1;
```

Comment:
Please add an explanation for +1.

Explanation can be found from:
https://vulkan-tutorial.com/Drawing_a_triangle/Presentation/Swap_chain